src/plugins/push/pushactivation.ts

```diff
@@ -0,0 +1,605 @@
+    machine.callDeactivatedCallback(null);
+    return NotActivated;
+  } else if (event instanceof GotPushDeviceDetails) {
+    const rest = machine.rest;
+    const device = rest.device as LocalDevice;
+
+    if (machine.registerCallback) {
+      machine.callCustomRegisterer(device, true);
+    } else {
+      const format = rest.options.useBinaryProtocol
+          ? machine.rest.Utils.Format.msgpack
+          : machine.rest.Utils.Format.json,
+        body = rest.DeviceDetails?.fromLocalDevice(device),
+        headers = machine.rest.Defaults.defaultPostHeaders(rest.options, { format }),
+        params = {};
+
+      if (rest.options.headers) machine.rest.Utils.mixin(headers, rest.options.headers);
+
+      if (rest.options.pushFullWait) machine.rest.Utils.mixin(params, { fullWait: 'true' });
+
+      const requestBody = machine.rest.Utils.encodeBody(body, rest._MsgPack, format);
+
+      machine.rest.Resource?.post(rest, '/push/deviceRegistrations', requestBody, headers, params, null, true)
+        .then((response) => {
+          const deviceDetails = response.unpacked
+            ? response.body
+            : rest.DeviceDetails?.fromResponseBody(response.body as any, rest._MsgPack, format);
+          machine.handleEvent(new GotDeviceRegistration(deviceDetails as DeviceRegistration));
+        })
+        .catch((err) => {
+          machine.handleEvent(new GettingDeviceRegistrationFailed(err as ErrorInfo));
+        });
+    }
+
+    return WaitingForDeviceRegistration;
+  } else if (event instanceof GettingPushDeviceDetailsFailed) {
+    machine.callActivatedCallback(event.reason);
+    return NotActivated;
+  }
+  return null;
+};
+
+const WaitingForDeviceRegistration = function (machine: ActivationStateMachine, event: ActivationEvent) {
+  if (event instanceof CalledActivate) {
+    return WaitingForDeviceRegistration;
+  } else if (event instanceof GotDeviceRegistration) {
+    const device = machine.rest.device as LocalDevice;
+    device.deviceIdentityToken = event.tokenDetails.token;
+    device.persist();
+    machine.callActivatedCallback(null);
+    return WaitingForNewPushDeviceDetails;
+  } else if (event instanceof GettingDeviceRegistrationFailed) {
+    machine.callActivatedCallback(event.reason);
+    return NotActivated;
+  }
+  return null;
+};
+
+const WaitingForNewPushDeviceDetails = function (machine: ActivationStateMachine, event: ActivationEvent) {
+  if (event instanceof CalledActivate) {
+    machine.callActivatedCallback(null);
+    return WaitingForNewPushDeviceDetails;
+  } else if (event instanceof CalledDeactivate) {
+    machine.deregister();
+    return WaitingForDeregistration(WaitingForNewPushDeviceDetails);
+  } else if (event instanceof GotPushDeviceDetails) {
+    machine.updateRegistration();
+    return WaitingForRegistrationSync;
+  }
+};
+
+const WaitingForRegistrationSync = function (machine: ActivationStateMachine, event: ActivationEvent) {
+  if (event instanceof CalledActivate) {
```

Comment:
is `unless the machine is in state WaitingForRegistrationSync as a result of a CalledActivate event` part of [RSH3e1](https://sdk.ably.com/builds/ably/specification/main/features/#RSH3e1) relevant here? Should we have an additional check for this?